app/multitenant/sqs_control_router.go

```diff
@@ -0,0 +1,330 @@
+package multitenant
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/sqs"
+	"golang.org/x/net/context"
+
+	"github.com/weaveworks/scope/app"
+	"github.com/weaveworks/scope/common/xfer"
+)
+
+var (
+	longPollTime = aws.Int64(10)
+	rpcTimeout   = time.Minute
+)
+
+// sqsControlRouter:
+// Creates a queue for every probe that connects to it, and a queue for
+// responses back to it.  When it recieves a request, posts it to the
+// probe queue.  When probe recieves a request, handles it and posts the
+// response back to the response queue.
+type sqsControlRouter struct {
+	service  *sqs.SQS
+	queueURL *string
+	userIDer UserIDer
+
+	mtx          sync.Mutex
+	responses    map[string]chan xfer.Response
+	probeWorkers map[int64]*probeWorker
+}
+
+type sqsRequestMessage struct {
+	ID               string
+	Request          xfer.Request
+	ResponseQueueURL string
+}
+
+type sqsResponseMessage struct {
+	ID       string
+	Response xfer.Response
+}
+
+// NewSQSControlRouter the harbinger of death
+func NewSQSControlRouter(url, region string, creds *credentials.Credentials, userIDer UserIDer) app.ControlRouter {
+	result := &sqsControlRouter{
+		service: sqs.New(session.New(aws.NewConfig().
+			WithEndpoint(url).
+			WithRegion(region).
+			WithCredentials(creds))),
+		queueURL:     nil,
+		userIDer:     userIDer,
+		responses:    map[string]chan xfer.Response{},
+		probeWorkers: map[int64]*probeWorker{},
+	}
+	go result.loop()
+	return result
+}
+
+func (cr *sqsControlRouter) Stop() error {
+	return nil
+}
+
+func (cr *sqsControlRouter) setQueueURL(url *string) {
+	cr.mtx.Lock()
+	defer cr.mtx.Unlock()
+	cr.queueURL = url
+}
+
+func (cr *sqsControlRouter) getQueueURL() *string {
+	cr.mtx.Lock()
+	defer cr.mtx.Unlock()
+	return cr.queueURL
+}
+
+func (cr *sqsControlRouter) getOrCreateQueue(name string) (*string, error) {
```

Comment:
> does the caller care if it is created here?

Not really, but I kinda like how the code reads when used - its very obvious what its doing.

> Also, is there a race between the get and create calls?

Good spot.  Luckily no; from http://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_CreateQueue.html:

"If you provide the name of an existing queue, along with the exact names and values of all the queue's attributes, CreateQueue returns the queue URL for the existing queue."

Which of course means I can just get rid of the GetQueueURL....